teps/0076-array-result-types.md

```diff
@@ -0,0 +1,321 @@
+* Be consistent in our treatment of arrays in Pipelines and Tasks and other types, e.g.
+  dictionaries (TEP-0075)
+
+### Non-Goals
+
+* Adding complete [JSONPath](https://tools.ietf.org/id/draft-goessner-dispatch-jsonpath-00.html)
+  syntax support
+* Adding support for nesting, e.g. array params where the items in the array are themselves arrays
+  (no reason we can't add that later but trying to keep it simple for now)
+
+### Use Cases
+
+1. Looping (i.e. providing values for an interface such as the task-loop custom task
+   [iterateParam](https://github.com/tektoncd/experimental/tree/main/task-loops#specifying-the-iteration-parameter)
+   (see the workaround added in [experimental#713](https://github.com/tektoncd/experimental/pull/713) which
+   splits a string). For example, in tektoncd/plumbing we have a directory of
+   [images](https://github.com/tektoncd/plumbing/tree/main/tekton/images). Each directory contains a Dockerfile -
+   in a Pipeline that builds all images in this repo, we would want to: 1) determine which Dockerfiles need to be built
+   and 2) for each of those Dockerfiles, build the image. Looping support would allow us to use a Task which builds
+   just one image (e.g. [the kaniko task](https://github.com/tektoncd/catalog/tree/main/task/kaniko/0.4)), and run it
+   for each Dockerfile, instead of having to the task to take multiple Dockerfiles as input. This missing piece of this
+   is supporting the Task which determines with Dockerfiles to build and provides them as an array (instead of
+   encoded in a string).
+
+   ```
+   filter-dockerfiles
+        |
+        | result: list of dockerfiles
+        v
+   (for each: filter-dockerfiles result)
+      kaniko build
+   ```
+
+## Requirements
+
+* Must be possible for a result array to be empty
+* Must be possible to use the array results of one Task in a Pipeline as the
+  param of another Task in the pipeline which has an array param
+* Must be possible to use one specific value in an array result of one Task
+  in a Pipeline as the param of another Task in the pipeline which has a string param
+  * If there is no value at the specified index, the Pipeline execution should fail
+    (or we may be able to use [TEP-0048](https://github.com/tektoncd/community/pull/240)
+    to specify a default value to use in that case instead)
+
+## Proposal
+
+Currently we treat the contents of
+[a Task's result file(s)](https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#emitting-results)
+as a string. The file is located at `$(results.resultName.path)` which expands to `/tekton/results/resultName`.
+In this proposal, we add the option to write to a json file instead, available at `$(results.resultName.jsonPath)`
+which expands to `/tekton/results/resultName.json`. The contents of this file would be treated as json and in addition
+to supporting a string value in this file, we sould support json arrays.
```

Comment:
Any reason for going `json` vs another format ?

Reply:
And potential issues ? (if someone write a complex json file, what do we do ?)

Reply:
@vdemeester do you have any other formats you'd suggest? off the top of my head all I can think of are:

* yaml (i dont think there is a syntax like jsonpath available for yaml? i.e. it'd be harder to expose via variable replacement - or we'd need to convert from yaml to json internally)
* protobuf (this would dramatically increase the difficulty in emitting structured results from tasks - either we'd see every Task including a step with protobuf tools in it or we'd need to include some by default)
* xml (seems more verbose than we'd need?)
* csv (would also be easy to use but afaik it wouldn't support more complex types like dicts? or we'd need to add some kind of syntax on top of the csv format to define more complex types)

as for why json I'd say because:
* it's human readable and so it's easy for folks to write json content without needing extra tools
* we've added json specific support to other parts of our APIs (e.g. [in triggers](https://github.com/tektoncd/triggers/blob/f2380d7d33666547ee7ed965730cf7dbc1aa5421/docs/triggerbindings.md#accessing-data-in-http-json-payloads))
* yaml and json are the two formats that k8s is using so I think it makes sense to use one of those

Reply:
> if someone write a complex json file, what do we do 

i think we'd simply bail out of evaluating the json as soon as it doesn't match the expected schema (which i think the "notes/caveats" tries to address but ill try to make it more clear) and also given the size limits, there's a limit on how complex the json file can be (maybe you have some specific scenario in mind tho?)

Reply:
| if someone write a complex json file, what do we do

ill add something to the testing section as well to make sure this is covered (a deeply nested dictionary is the most complex scenario i can think of?)